### PR TITLE
Add deepseek_v3 fused gate

### PR DESCRIPTION
**Pull request by NovTi**
Add deepseek v3 fused gate module

### CODE REVIEW

File: sgl-kernel/tests/test_deepseek_fused_gate.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    torch.manual_seed(seq_length)
+
+    tensor = torch.rand((seq_length, 256)).to(torch.bfloat16).cuda()
+    scores = tensor.clone()
+    bias = torch.rand(256).to(torch.bfloat16).cuda()
+
+    # Your module under test
+    output, indices_my = deepseekv3_fused_gate(tensor, bias, seq_length)
+
+    ###### Reference Implementation ######
```
Comment (BBuf) [inline sgl-kernel/tests/test_deepseek_fused_gate.py]: Please refactor this code into a standalone function, which can be directly used from https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/moe/topk.py#L111-L147.

File: sgl-kernel/tests/test_deepseek_fused_gate.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    scores = (scores * mask.unsqueeze(-1)).flatten(1)
+    indices = torch.topk(scores, 8, dim=-1)[1]
+    weights = original_scores.gather(1, indices)
+
+    weights /= weights.sum(dim=-1, keepdim=True)
+    weights *= 2.5
+    output_ref = weights.type_as(scores)
+
+    # Assertions
+    output_check = torch.allclose(output_ref.sort()[0], output.sort()[0], rtol=1e-04, atol=1e-05)
```
Comment (BBuf) [inline sgl-kernel/tests/test_deepseek_fused_gate.py]: Why not directly compare output and output_ref instead of sorting them?

File: sgl-kernel/tests/test_deepseek_fused_gate.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import torch
+import pytest
+from tqdm import tqdm
+from sgl_kernel import deepseekv3_fused_gate
+
+
+@pytest.mark.parametrize("seq_length", range(1, 20000))
```
Comment (BBuf) [inline sgl-kernel/tests/test_deepseek_fused_gate.py:8]: Can you add a benchmark script? Maybe refer to https://github.com/sgl-project/sglang/tree/main/sgl-kernel/benchmark

File: sgl-kernel/src/sgl-kernel/__init__.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
     bmm_fp8,
     custom_dispose,
     custom_reduce,
+    deepseekv3_fused_gate,
```
Comment (BBuf) [inline sgl-kernel/src/sgl-kernel/__init__.py:6]: It seems more appropriate to name it `deepseekv3_fused_gate` here, as models from the deepseek series can all go through this `gate` function.

File: sgl-kernel/tests/test_deepseek_fused_gate.py (hunk 1 above)
Comment (NovTi) [reply sgl-kernel/tests/test_deepseek_fused_gate.py]: Do you mean I separate the reference implementation into a standalone function?

Comment (BBuf) [reply sgl-kernel/tests/test_deepseek_fused_gate.py]: Yeah.

Comment (NovTi) [reply sgl-kernel/tests/test_deepseek_fused_gate.py]: Got it, I will do that

File: sgl-kernel/tests/test_deepseek_fused_gate.py (hunk 2 above)
Comment (NovTi) [reply sgl-kernel/tests/test_deepseek_fused_gate.py]: This is weird, kernel sometimes will output exact same output but in a different order. I checked the following steps and the output order does not matter so I used this way to do the unit test, is this ok?

File: sgl-kernel/src/sgl-kernel/__init__.py (hunk 4 above)
Comment (NovTi) [reply sgl-kernel/src/sgl-kernel/__init__.py:6]: This is not a generalized kernel, it only works for deepseek v3 671b model

File: sgl-kernel/tests/test_deepseek_fused_gate.py (hunk 3 above)
Comment (NovTi) [reply sgl-kernel/tests/test_deepseek_fused_gate.py:8]: Sure

File: sgl-kernel/tests/test_deepseek_fused_gate.py (hunk 2 above)
Comment (BBuf) [reply sgl-kernel/tests/test_deepseek_fused_gate.py]: We need to determine at which specific step of the fused kernel this inconsistency in order occurs. Additionally, we need to clarify whether running the PyTorch implementation twice with the same input would result in inconsistent output orders. Finally, if you believe that the current order inconsistency does not affect the fused MoE accuracy, you need to provide an end-to-end result, such as running the GSM8K test with the DeepSeek V3 model.

图片

File: sgl-kernel/src/sgl-kernel/__init__.py (hunk 4 above)
Comment (BBuf) [reply sgl-kernel/src/sgl-kernel/__init__.py:6]: I see, thanks.

File: sgl-kernel/tests/test_deepseek_fused_gate.py (hunk 2 above)
Comment (NovTi) [reply sgl-kernel/tests/test_deepseek_fused_gate.py]: I see, I will check the inconsistency inside the kernel. I cannot run e2e test on my server, Yineng will help me do the test

File: sgl-kernel/src/sgl-kernel/csrc/deepseek_fused_gate.cu [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    int64_t const num_blocks = (num_warps + WARPS_PER_CTA - 1) / WARPS_PER_CTA;
+
+    const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+    dim3 block_dim(32, WARPS_PER_CTA);
+    deepseekv3_fused_gate_kernel<VPT, NUM_EXPERTS, THREADS_PER_ROW, ROWS_PER_WARP, ROWS_PER_CTA, WARPS_PER_CTA><<<num_blocks, block_dim, 0, stream>>>(
+        input.data_ptr(), bias.data_ptr(), output.data_ptr(), indices.data_ptr<int64_t>(), num_rows, k, route_scale
+    );
+
+    CHECK_CUDA_SUCCESS(cudaDeviceSynchronize());
```
Comment (BBuf) [inline sgl-kernel/src/sgl-kernel/csrc/deepseek_fused_gate.cu]: Synchronization is not allowed in CUDA kernel's host code, as it will cause CUDA graphs to crash. Can you remove it?

Comment (NovTi) [reply sgl-kernel/src/sgl-kernel/csrc/deepseek_fused_gate.cu]: Thanks, I will update these

File: sgl-kernel/src/sgl-kernel/__init__.py (hunk 4 above)
Comment (zhyncs) [reply sgl-kernel/src/sgl-kernel/__init__.py:6]: I think it also works for DeepSeek V2 VL

File: sgl-kernel/src/sgl-kernel/csrc/deepseek_fused_gate.cu [hunk 6]
```diff
@@ -0,0 +1,219 @@
+#include <cfloat>
```
Comment (BBuf) [inline sgl-kernel/src/sgl-kernel/csrc/deepseek_fused_gate.cu:1]: Please add `Adapted from https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/kernels/mixtureOfExperts/moe_kernels.cu#L231`

**Comment by BBuf** [comment]
In TensorRT-LLM, the fused MoE module, in addition to the `fused_gate` here, also includes the `trt_moe_expand_and_permute`, the CUTLASS `grouped GEMM`, and the `trt_moe_unpermute_and_reduce` processes. Compared to the MoE implemented in Triton, the advantage of TensorRT-LLM's approach is that it does not require padding, which saves some computational overhead, and the CUTLASS implementation may have greater performance potential, especially on Hopper architecture. I conducted an experiment at https://github.com/sgl-project/sglang/tree/bbuf_tmp, where I successfully connected `trt_moe_expand_and_permute`, `trt_moe_unpermute_and_reduce`, and FlashInfer's `grouped GEMM` in sgl-kernel to run correctness comparisons with the Triton fused MoE operator in the case of `bfloat16` dtype. However, it seems that the current performance is still significantly worse than Triton's. This could be due to performance issues with FlashInfer's grouped GEMM on specific shapes. Additionally, FlashInfer's GEMM does not currently support scaled FP8 or INT8 GEMM. If anyone is interested, we can discuss whether to directly integrate TensorRT's fused MoE as a backend into sglang or to use FlashInfer's approach, which would require a customization of FlashInfer for grouped GEMM. cc @zhyncs

**Comment by zhyncs** [comment]
> directly integrate TensorRT's fused MoE as a backend into sglang

sounds good @BBuf

**Comment by BBuf** [comment]
> > directly integrate TensorRT's fused MoE as a backend into sglang
> 
> sounds good @BBuf

Yeah, I can have a try.

**Comment by yiakwy-xpu-ml-framework-team** [comment]
@NovTi are you still working on the PR ? Have you reference to any other implementation open sourced ?

File: sgl-kernel/src/sgl-kernel/csrc/deepseek_fused_gate.cu (hunk 6 above)
Comment (yiakwy-xpu-ml-framework-team) [reply sgl-kernel/src/sgl-kernel/csrc/deepseek_fused_gate.cu:1]: Note the codes has been removed from v0.16.0 into static lib (closed source) in v0.17.0
